br/pkg/lightning/backend/external/byte_reader.go

```diff
@@ -63,69 +80,118 @@ func openStoreReaderAndSeek(
 	return storageReader, nil
 }
 
-// newByteReader wraps readNBytes functionality to storageReader.
+// newByteReader wraps readNBytes functionality to storageReader. If store and
+// filename are also given, this reader can use switchConcurrentMode to switch to
+// concurrent reading mode.
 func newByteReader(
 	ctx context.Context,
 	storageReader storage.ExternalFileReader,
 	bufSize int,
-	st storage.ExternalStorage,
-	name string,
-	defaultUseConcurrency bool,
 ) (r *byteReader, err error) {
 	defer func() {
 		if err != nil && r != nil {
 			_ = r.Close()
 		}
 	}()
-	fileSize, err := storageReader.GetFileSize()
-	if err != nil {
-		return nil, err
-	}
-	conReader, err := newSingeFileReader(ctx, st, name, fileSize, 8, ConcurrentReaderBufferSize)
-	if err != nil {
-		return nil, err
-	}
 	r = &byteReader{
 		ctx:           ctx,
 		storageReader: storageReader,
-		buf:           make([]byte, bufSize),
-		bufOffset:     0,
-		conReader:     conReader,
+		smallBuf:      make([]byte, bufSize),
+		curBufOffset:  0,
 	}
-	r.switchReaderMode(defaultUseConcurrency)
+	r.curBuf = r.smallBuf
+	r.logger = logutil.Logger(r.ctx)
 	return r, r.reload()
 }
 
-// switchReaderMode switches to concurrent reader.
-func (r *byteReader) switchReaderMode(useConcurrent bool) {
-	r.useConcurrentReader.Store(useConcurrent)
+func (r *byteReader) enableConcurrentRead(
+	store storage.ExternalStorage,
+	filename string,
+	concurrency int,
+	bufSizePerConc int,
+	bufferPool *membuf.Buffer,
+) {
+	r.concurrentReader.store = store
+	r.concurrentReader.filename = filename
+	r.concurrentReader.concurrency = concurrency
+	r.concurrentReader.bufSizePerConc = bufSizePerConc
+	r.concurrentReader.largeBufferPool = bufferPool
 }
 
-func (r *byteReader) switchToConcurrentReaderImpl() error {
-	if r.conReader == nil {
-		return errors.New("can't use the concurrent mode because reader is not initialized correctly")
+// switchConcurrentMode is used to help implement sortedReader.switchConcurrentMode.
+// See the comment of the interface.
+func (r *byteReader) switchConcurrentMode(useConcurrent bool) error {
+	readerFields := &r.concurrentReader
+	if readerFields.store == nil {
+		r.logger.Warn("concurrent reader is not enabled, skip switching")
+		// caller don't need to care about it.
+		return nil
+	}
+	// need to set it before reload()
+	readerFields.expected = useConcurrent
+	// concurrent reader will be lazily initialized when reload()
+	if useConcurrent {
+		return nil
+	}
+
+	// no change
+	if !readerFields.now && !useConcurrent {
```

Comment:
`!useConcurrent` always is true.

Reply:
https://github.com/pingcap/tidb/pull/47239/commits/e8773f5304fd46e87e21a51ff8bacb28b440e4e5